File: frontend/ui-architecture-playground/mvp-vanillajs/js/presenters/users-presenter.js
// ============================================================================
// UsersPresenter - презентер экрана пользователей. JS-порт UsersPresenter.
// Связь с главным экраном - колбэк onChanged (аналог события Changed ->
// boardPresenter.externalRefresh() в композиционном корне app.js).
// ============================================================================

'use strict';

class UsersPresenter {
    constructor(view, store, onChanged) {
        this.view = view;
        this.store = store;
        this.onChanged = onChanged;
    }

    run() {
        this.view.bindHandlers({
            addUser: () => this.addUser(),
            deleteUser: id => this.deleteUser(id),
            close: () => this.onChanged?.()          // вернулись на доску - она перечитает данные
        });
    }

    open() {
        if (!this.view.isOpen) {
            this.view.open();
            this.reload();
        }
    }

    reload() {
        this.view.showUsers(this.store.users.map(u => ({
            id: u.id,
            name: u.name,
            taskCount: this.store.countTasksAssignedTo(u.id)
        })));
    }

    addUser() {
        const name = this.view.newUserName.trim();
        if (!name) {
            this.view.flash('Введите имя пользователя');
            return;
        }
        this.store.addUser(name);
        this.view.clearNewUserName();                // защита от дубликатов по повторному клику
        this.onChanged?.();                          // у доски появятся новые опции фильтров/назначения
        this.reload();
        this.view.flash(`Пользователь «${name}» добавлен`);
    }

    async deleteUser(userId) {
        const user = this.store.users.find(u => u.id === userId);
        if (!user) return;

        const count = this.store.countTasksAssignedTo(userId);

        // gherkin: если задач нет - просто подтвердить; иначе предложить переназначение.
        const choice = await this.view.askDeleteUser(
            user.name,
            this.store.users.filter(u => u.id !== userId)
                .map(u => ({ id: u.id, label: u.name })),
            count);
        if (!choice.confirmed) return;

        this.store.deleteUser(userId, choice.reassignTo);
        const targetName = choice.reassignTo
            ? this.store.users.find(u => u.id === choice.reassignTo)?.name
            : null;

        // сначала перерисовка, потом flash: reload() пересобирает панель
        // и стёр бы сообщение, проставленное до неё
        this.onChanged?.();                          // карточки на доске сменили исполнителя
        this.reload();
        this.view.flash(targetName
            ? `Пользователь удалён; задачи переназначены на ${targetName}`
            : 'Пользователь удалён; его задачи остались нераспределёнными');
    }
}

File: frontend/ui-architecture-playground/mvp-winforms/BoardApp/Presenters/UsersPresenter.cs
// ============================================================================
// UsersPresenter - презентер экрана пользователей.
// Вторая пара View<->Presenter: показывает, что в MVP "экран = пара".
// Связь с главным экраном - НЕ прямые ссылки, а колбэк onChanged,
// который Program.cs подключает к BoardPresenter.ExternalRefresh.
// ============================================================================

using BoardApp.Core;
using BoardApp.Views.Contracts;

namespace BoardApp.Presenters;

public sealed class UsersPresenter(IUsersView view, IBoardStore store)
{
    /// <summary>Вызывается после каждой мутации - композиционный корень уведомит другие экраны.</summary>
    public event Action? Changed;

    public void Run()
    {
        view.AddUserRequested += (_, _) => AddUser();
        view.DeleteUserRequested += (_, e) => DeleteUser(e.Id);
        Reload();
    }

    private void Reload() =>
        view.ShowUsers(store.Users
            .Select(u => new UserRow(u.Id, u.Name, store.CountTasksAssignedTo(u.Id)))
            .ToList());

    private void AddUser()
    {
        var name = view.NewUserName.Trim();
        if (name.Length == 0)
        {
            view.ShowFlash("Введите имя пользователя");
            return;
        }
        store.AddUser(name);
        view.ShowFlash($"Пользователь \"{name}\" добавлен");
        Changed?.Invoke();                                          // у главного экрана появятся новые опции фильтров/назначения
        Reload();                                                   // счётчики задач могли не измениться, но список - да
    }

    private void DeleteUser(Guid userId)
    {
        var user = store.Users.FirstOrDefault(u => u.Id == userId);
        if (user is null) return;

        var count = store.CountTasksAssignedTo(userId);

        // gherkin: если задач нет - просто подтвердить; иначе предложить переназначение.
        // Весь диалог - метод КОНТРАКТА view; presenter лишь интерпретирует результат.
        var choice = view.AskDeleteUser(
            user.Name,
            store.Users.Where(u => u.Id != userId)
                .Select(u => new OptionVm(u.Id, u.Name)).ToList(),
            count);
        if (!choice.Confirmed) return;

        store.DeleteUser(userId, choice.ReassignTo);
        view.ShowFlash(choice.ReassignTo is { } target
            ? $"Пользователь удалён; задачи переназначены на {store.Users.First(u => u.Id == target).Name}"
            : "Пользователь удалён; его задачи остались нераспределёнными");

        Changed?.Invoke();                                          // карточки на доске сменили исполнителя
        Reload();
    }
}
